File: internal/model/whatsapp.go
package model

import "time"

// WhatsAppCallDirection enumerates the direction of a WhatsApp call.
type WhatsAppCallDirection string

// Supported call directions.
const (
	WhatsAppCallDirectionInbound  WhatsAppCallDirection = "inbound"
	WhatsAppCallDirectionOutbound WhatsAppCallDirection = "outbound"
)

// WhatsAppCallState enumerates the lifecycle states of a WhatsApp call.
type WhatsAppCallState string

// Supported call states.
const (
	WhatsAppCallStateRinging   WhatsAppCallState = "ringing"
	WhatsAppCallStateConnected WhatsAppCallState = "connected"
	WhatsAppCallStateEnded     WhatsAppCallState = "ended"
)

// WhatsAppPhoneNumber represents a provisioned WhatsApp Business phone number.
type WhatsAppPhoneNumber struct {
	ID          string    `json:"id"`
	OrgID       string    `json:"org_id"`
	PhoneNumber string    `json:"phone_number"`
	DisplayName string    `json:"display_name"`
	WABAID      string    `json:"waba_id"`
	Verified    bool      `json:"verified"`
	CreatedAt   time.Time `json:"created_at"`
	UpdatedAt   time.Time `json:"updated_at"`
}

// WhatsAppCall represents a single WhatsApp Business API call record.
type WhatsAppCall struct {
	ID              string                `json:"id"`
	OrgID           string                `json:"org_id"`
	CallID          string                `json:"call_id"`
	PhoneNumberID   string                `json:"phone_number_id"`
	Direction       WhatsAppCallDirection `json:"direction"`
	State           WhatsAppCallState     `json:"state"`
	Caller          string                `json:"caller"`
	Callee          string                `json:"callee"`
	From            string                `json:"from,omitempty"`
	To              string                `json:"to,omitempty"`
	SDPOffer        *string               `json:"sdp_offer,omitempty"   db:"sdp_offer"`
	SDPAnswer       *string               `json:"sdp_answer,omitempty"  db:"sdp_answer"`
	StartedAt       *time.Time            `json:"started_at,omitempty"`
	EndedAt         *time.Time            `json:"ended_at,omitempty"`
	DurationSeconds *int                  `json:"duration_seconds,omitempty"`
	CreatedAt       time.Time             `json:"created_at"`
	UpdatedAt       time.Time             `json:"updated_at"`
}

// --- Request / Response types ---

// CreateWhatsAppPhoneNumberRequest is the payload for POST /whatsapp/phone-numbers.
type CreateWhatsAppPhoneNumberRequest struct {
	PhoneNumber string `json:"phone_number" binding:"required,max=20"`
	DisplayName string `json:"display_name" binding:"omitempty,max=255"`
	WABAID      string `json:"waba_id" binding:"required,max=64"`
}

// UpdateWhatsAppPhoneNumberRequest is the payload for PUT /whatsapp/phone-numbers/:id.
type UpdateWhatsAppPhoneNumberRequest struct {
	DisplayName *string `json:"display_name,omitempty" binding:"omitempty,max=255"`
	Verified    *bool   `json:"verified,omitempty"`
}

// WhatsAppPhoneNumberListResponse is the paginated list of phone numbers.
type WhatsAppPhoneNumberListResponse struct {
	PhoneNumbers []WhatsAppPhoneNumber `json:"phone_numbers"`
	Total        int                   `json:"total"`
	Limit        int                   `json:"limit"`
	Offset       int                   `json:"offset"`
}

// InitiateWhatsAppCallRequest is the payload for POST /whatsapp/calls.
type InitiateWhatsAppCallRequest struct {
	PhoneNumberID string `json:"phone_number_id" binding:"required,uuid"`
	Callee        string `json:"callee" binding:"required,max=20"`
}

// UpdateWhatsAppCallStateRequest is the payload for PATCH /whatsapp/calls/:id.
type UpdateWhatsAppCallStateRequest struct {
	State WhatsAppCallState `json:"state" binding:"required"`
}

// WhatsAppCallListResponse is the paginated list of calls.
type WhatsAppCallListResponse struct {
	Calls  []WhatsAppCall `json:"calls"`
	Total  int            `json:"total"`
	Limit  int            `json:"limit"`
	Offset int            `json:"offset"`
}

// WhatsAppCallResponse wraps a single WhatsAppCall for JSON API responses.
type WhatsAppCallResponse struct {
	Call WhatsAppCall `json:"call"`
}

// SendSDPAnswerRequest is the payload for POST /whatsapp/calls/:call_id/answer.
type SendSDPAnswerRequest struct {
	SDPAnswer string `json:"sdp_answer" binding:"required"`
}

// --- Webhook payload types (Meta Graph API) ---

// WhatsAppWebhookPayload is the top-level structure of a Meta webhook delivery.
type WhatsAppWebhookPayload struct {
	Object string                 `json:"object"`
	Entry  []WhatsAppWebhookEntry `json:"entry"`
}

// WhatsAppWebhookEntry represents a single entry in the webhook payload.
type WhatsAppWebhookEntry struct {
	ID      string                  `json:"id"`
	Changes []WhatsAppWebhookChange `json:"changes"`
}

// WhatsAppWebhookChange represents a single change within an entry.
type WhatsAppWebhookChange struct {
	Field string                     `json:"field"`
	Value WhatsAppWebhookChangeValue `json:"value"`
}

// WhatsAppWebhookChangeValue holds the call-related data from Meta's webhook.
type WhatsAppWebhookChangeValue struct {
	MessagingProduct string `json:"messaging_product"`
	CallID           string `json:"call_id,omitempty"`
	From             string `json:"from,omitempty"`
	To               string `json:"to,omitempty"`
	Status           string `json:"status,omitempty"`
	Timestamp        string `json:"timestamp,omitempty"`
}

File: frontend/src/api/whatsapp.ts
// --- Phone number types ---

export interface WhatsAppPhoneNumber {
  id: string
  org_id: string
  phone_number: string
  display_name: string
  waba_id: string
  verified: boolean
  created_at: string
  updated_at: string
}

export interface WhatsAppPhoneNumberListResponse {
  phone_numbers: WhatsAppPhoneNumber[]
  total: number
  limit: number
  offset: number
}

export interface CreatePhoneNumberRequest {
  phone_number: string
  display_name?: string
  waba_id: string
}

// --- Call types ---

export type CallDirection = 'inbound' | 'outbound'
export type CallState = 'ringing' | 'connected' | 'ended'

export interface WhatsAppCall {
  id: string
  org_id: string
  call_id: string
  phone_number_id: string
  direction: CallDirection
  state: CallState
  caller: string
  callee: string
  from?: string
  to?: string
  started_at?: string
  ended_at?: string
  duration_seconds?: number
  created_at: string
  updated_at: string
}

export interface WhatsAppCallListResponse {
  calls: WhatsAppCall[]
  total: number
  limit: number
  offset: number
}

export interface InitiateCallRequest {
  phone_number_id: string
  callee: string
}

// --- Bridge types ---

export type BridgeState = 'initializing' | 'active' | 'failed' | 'closed'

export interface WhatsAppBridge {
  id: string
  org_id: string
  call_id: string
  livekit_room: string
  bridge_state: BridgeState
  voice_session_id?: string
  metadata?: string
  created_at: string
  updated_at: string
  closed_at?: string
}

export interface BridgeStatusResponse {
  bridge: WhatsAppBridge
}

// --- API helpers ---

const API_BASE = import.meta.env.VITE_API_BASE_URL ?? '/api/v1'

async function authFetch(path: string, init?: RequestInit): Promise<Response> {
  return fetch(API_BASE + path, {
    ...init,
    credentials: 'include',
    headers: {
      'Content-Type': 'application/json',
      ...init?.headers,
    },
  })
}

// --- Phone number endpoints ---

export async function listPhoneNumbers(
  orgId: string,
  offset = 0,
  limit = 20,
): Promise<WhatsAppPhoneNumberListResponse> {
  const res = await authFetch(
    `/orgs/${orgId}/whatsapp/phone-numbers?offset=${offset}&limit=${limit}`,
  )
  if (!res.ok) throw new Error(`listPhoneNumbers failed: ${res.status}`)
  return res.json()
}

export async function createPhoneNumber(
  orgId: string,
  req: CreatePhoneNumberRequest,
): Promise<WhatsAppPhoneNumber> {
  const res = await authFetch(`/orgs/${orgId}/whatsapp/phone-numbers`, {
    method: 'POST',
    body: JSON.stringify(req),
  })
  if (!res.ok) throw new Error(`createPhoneNumber failed: ${res.status}`)
  return res.json()
}

export async function deletePhoneNumber(orgId: string, phoneId: string): Promise<void> {
  const res = await authFetch(`/orgs/${orgId}/whatsapp/phone-numbers/${phoneId}`, {
    method: 'DELETE',
  })
  if (!res.ok) throw new Error(`deletePhoneNumber failed: ${res.status}`)
}

// --- Call endpoints ---

export async function listCalls(
  orgId: string,
  offset = 0,
  limit = 20,
): Promise<WhatsAppCallListResponse> {
  const res = await authFetch(`/orgs/${orgId}/whatsapp/calls?offset=${offset}&limit=${limit}`)
  if (!res.ok) throw new Error(`listCalls failed: ${res.status}`)
  return res.json()
}

export async function getCall(orgId: string, callId: string): Promise<WhatsAppCall> {
  const res = await authFetch(`/orgs/${orgId}/whatsapp/calls/${callId}`)
  if (!res.ok) throw new Error(`getCall failed: ${res.status}`)
  return res.json()
}

export async function initiateCall(
  orgId: string,
  req: InitiateCallRequest,
): Promise<WhatsAppCall> {
  const res = await authFetch(`/orgs/${orgId}/whatsapp/calls`, {
    method: 'POST',
    body: JSON.stringify(req),
  })
  if (!res.ok) throw new Error(`initiateCall failed: ${res.status}`)
  return res.json()
}

export async function updateCallState(
  orgId: string,
  callId: string,
  state: CallState,
): Promise<WhatsAppCall> {
  const res = await authFetch(`/orgs/${orgId}/whatsapp/calls/${callId}`, {
    method: 'PATCH',
    body: JSON.stringify({ state }),
  })
  if (!res.ok) throw new Error(`updateCallState failed: ${res.status}`)
  return res.json()
}

// --- Bridge endpoints ---

export async function getBridgeStatus(
  orgId: string,
  callId: string,
): Promise<BridgeStatusResponse> {
  const res = await authFetch(`/orgs/${orgId}/whatsapp/calls/${callId}/bridge`)
  if (!res.ok) throw new Error(`getBridgeStatus failed: ${res.status}`)
  return res.json()
}

export async function listActiveBridges(orgId: string): Promise<WhatsAppBridge[]> {
  const res = await authFetch(`/orgs/${orgId}/whatsapp/bridges`)
  if (!res.ok) throw new Error(`listActiveBridges failed: ${res.status}`)
  return res.json()
}
